### ISSUE TITLE
Empty SG chain after power cycle 

### ISSUE THREAD

**Issue by DaveLangridge** [Metaswitch/calico#645]
The test creates some VMs, does some pings, then power cycles one of the compute nodes. Then it does some more ping tests. The test fails more times than it passes (2:1 at a guess). A similar test which reboots the host passes most/all of the time.

When it fails, the test script can’t ssh into a VM on the host that’s been cycled. Looking at the iptables, the structure seems to be present, but most of the SG stuff is missing:

Chain felix-from-cd250fed-c7 (1 references)
pkts bytes target     prot opt in     out     source               destination
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate INVALID
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    2   630 RETURN     udp  --  *      *       0.0.0.0/0            0.0.0.0/0            udp spt:68 dpt:67
   11   924 MARK       all  --  *      *       0.0.0.0/0            0.0.0.0/0            MARK and 0x0
   11   924 felix-p-_a0438770e0e8efc-o  all  --  *      *       10.28.1.171          0.0.0.0/0            MAC FA:16:3E:20:5A:19
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match ! 0x1/0x1 /* No mark means profile accepted packet */
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* Default DROP if no match (endpoint cd250fed-c73e-41b9-8dd2-e7f01512cbe7): */

Chain felix-p-_a0438770e0e8efc-i (1 references)
pkts bytes target     prot opt in     out     source               destination
   59  4584 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* WARNING Missing chain DROP: */

Chain felix-p-_a0438770e0e8efc-o (1 references)
pkts bytes target     prot opt in     out     source               destination
   11   924 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* WARNING Missing chain DROP: */

Chain felix-to-cd250fed-c7 (1 references)
pkts bytes target     prot opt in     out     source               destination
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate INVALID
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
   59  4584 MARK       all  --  *      *       0.0.0.0/0            0.0.0.0/0            MARK and 0x0
   59  4584 felix-p-_a0438770e0e8efc-i  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match ! 0x1/0x1 /* No mark means profile accepted packet */
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* Endpoint cd250fed-c73e-41b9-8dd2-e7f01512cbe7: */

Which should be:

Chain felix-from-cd250fed-c7 (1 references)
pkts bytes target     prot opt in     out     source               destination
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate INVALID
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 RETURN     udp  --  *      *       0.0.0.0/0            0.0.0.0/0            udp spt:68 dpt:67
    0     0 MARK       all  --  *      *       0.0.0.0/0            0.0.0.0/0            MARK and 0x0
    0     0 felix-p-_a0438770e0e8efc-o  all  --  *      *       10.28.1.171          0.0.0.0/0            MAC FA:16:3E:20:5A:19
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match ! 0x1/0x1 /* No mark means profile accepted packet */
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* Default DROP if no match (endpoint cd250fed-c73e-41b9-8dd2-e7f01512cbe7): */

Chain felix-p-_a0438770e0e8efc-i (1 references)
pkts bytes target     prot opt in     out     source               destination
    0     0 RETURN     icmp --  *      *       0.0.0.0/0            0.0.0.0/0            match-set felix-v4-_a0438770e0e8efc src
    0     0 RETURN     tcp  --  *      *       172.18.203.223       0.0.0.0/0            multiport dports 22
    0     0 MARK       all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* Mark as not matched */ MARK set 0x1

Chain felix-p-_a0438770e0e8efc-o (1 references)
pkts bytes target     prot opt in     out     source               destination
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 MARK       all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* Mark as not matched */ MARK set 0x1

Chain felix-to-cd250fed-c7 (1 references)
pkts bytes target     prot opt in     out     source               destination
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate INVALID
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
    0     0 MARK       all  --  *      *       0.0.0.0/0            0.0.0.0/0            MARK and 0x0
    0     0 felix-p-_a0438770e0e8efc-i  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match ! 0x1/0x1 /* No mark means profile accepted packet */
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* Endpoint cd250fed-c73e-41b9-8dd2-e7f01512cbe7: */

The power cycle occurred at ~10:56. There are some logs like

2015-06-10 10:56:01,232 [INFO][1768/18] calico.felix.profilerules 203: Can't program rules 97dd63d7-4853-4aaa-970f-b1f46d09a670 yet, waiting on ipsets

But later it seems to work:

2015-06-10 10:56:01,294 [INFO][1768/5] calico.felix.fiptables 189: Iptables update: {'felix-p-_a0438770e0e8efc-i': ['--append felix-p-_a0438770e0e8efc-i --protocol tcp --source fec0::250:56ff:fe92:e7/128 --match multiport --destination-ports 22 --jump RETURN', '--append felix-p-_a0438770e0e8efc-i --match comment --comment "Mark as not matched" --jump MARK --set-mark 1'], 'felix-p-_a0438770e0e8efc-o': ['--append felix-p-_a0438770e0e8efc-o --destination ::/0 --jump RETURN', '--append felix-p-_a0438770e0e8efc-o --match comment --comment "Mark as not matched" --jump MARK --set-mark 1']}
2015-06-10 10:56:01,294 [INFO][1768/5] calico.felix.fiptables 190: Iptables deps: {}
2015-06-10 10:56:01,296 [INFO][1768/17] calico.felix.endpoint 393: Endpoint EndpointId<calico-st03/openstack/b10597c5-3b94-4343-a70a-194bf4d4b51f/cd250fed-c73e-41b9-8dd2-e7f01512cbe7> acquired all required profile references
2015-06-10 10:56:01,302 [INFO][1768/4] calico.felix.fiptables 336: IptablesUpdater<queue_len=1,live=True,msg=None> Successfully processed iptables updates.
2015-06-10 10:56:01,303 [INFO][1768/4] calico.felix.fiptables 189: Iptables update: {'felix-p-_a0438770e0e8efc-i': ['--append felix-p-_a0438770e0e8efc-i --protocol icmp --match set --match-set felix-v4-_a0438770e0e8efc src --jump RETURN', '--append felix-p-_a0438770e0e8efc-i --protocol tcp --source 172.18.203.223/32 --match multiport --destination-ports 22 --jump RETURN', '--append felix-p-_a0438770e0e8efc-i --match comment --comment "Mark as not matched" --jump MARK --set-mark 1'], 'felix-p-_a0438770e0e8efc-o': ['--append felix-p-_a0438770e0e8efc-o --destination 0.0.0.0/0 --jump RETURN', '--append felix-p-_a0438770e0e8efc-o --match comment --comment "Mark as not matched" --jump MARK --set-mark 1']}
2015-06-10 10:56:01,303 [INFO][1768/4] calico.felix.fiptables 190: Iptables deps: {}
2015-06-10 10:56:01,304 [INFO][1768/15] calico.felix.endpoint 393: Endpoint EndpointId<calico-st03/openstack/b10597c5-3b94-4343-a70a-194bf4d4b51f/cd250fed-c73e-41b9-8dd2-e7f01512cbe7> acquired all required profile references
2015-06-10 10:56:01,324 [INFO][1768/5] calico.felix.fiptables 336: IptablesUpdater<queue_len=0,live=True,msg=None> Successfully processed iptables updates.
2015-06-10 10:56:01,339 [INFO][1768/4] calico.felix.fiptables 336: IptablesUpdater<queue_len=0,live=True,msg=None> Successfully processed iptables updates.

But later (~12:10) , I restarted calico-felix and my rules came back. The ‘Iptables update’ logs look similar to the above.

Full diags here: https://transfer.sh/vNLJM/controller-diags.tar.xz

**Comment by DaveLangridge** [comment]
Marking as 0.24 for now. This isn't going to occur ofter (power cycle, and not 100% of those), but when it does occur it is silent, and doesn't recover.

**Comment by fasaxc** [comment]
This is odd.  The rules actor gets the ipsets that it depends on and then programs its chains:
```
$ grep  1768/18 felix.log
2015-06-10 10:56:01,232 [INFO][1768/18] calico.felix.profilerules 203: Can't program rules 97dd63d7-4853-4aaa-970f-b1f46d09a670 yet, waiting on ipsets
2015-06-10 10:56:01,283 [INFO][1768/18] calico.felix.profilerules 146: All required ipsets acquired.
2015-06-10 10:56:01,284 [INFO][1768/18] calico.felix.profilerules 191: Ready to program rules for 97dd63d7-4853-4aaa-970f-b1f46d09a670
2015-06-10 10:56:01,284 [INFO][1768/18] calico.felix.profilerules 209: ProfileRules<queue_len=0,live=True,msg=None> Programming iptables with our chains.
2015-06-10 10:56:01,284 [INFO][1768/18] calico.felix.profilerules 214: Updating inbound chain 'felix-p-_a0438770e0e8efc-i' for profile 97dd63d7-4853-4aaa-970f-b1f46d09a670
2015-06-10 10:56:01,284 [INFO][1768/18] calico.felix.profilerules 214: Updating outbound chain 'felix-p-_a0438770e0e8efc-o' for profile 97dd63d7-4853-4aaa-970f-b1f46d09a670
```
Then the IpsetUpdater receives the update, logs it out and says it succeeded:
```
2015-06-10 10:56:01,294 [INFO][1768/5] calico.felix.fiptables 189: Iptables update: {'felix-p-_a0438770e0e8efc-i': ['--append felix-p-_a0438770e0e8efc-i --protocol tcp --source fec0::250:56ff:fe92:e7/128 --match multiport --destination-ports 22 --jump RETURN', '--append felix-p-_a0438770e0e8efc-i --match comment --comment "Mark as not matched" --jump MARK --set-mark 1'], 'felix-p-_a0438770e0e8efc-o': ['--append felix-p-_a0438770e0e8efc-o --destination ::/0 --jump RETURN', '--append felix-p-_a0438770e0e8efc-o --match comment --comment "Mark as not matched" --jump MARK --set-mark 1']}
2015-06-10 10:56:01,294 [INFO][1768/5] calico.felix.fiptables 190: Iptables deps: {}
2015-06-10 10:56:01,324 [INFO][1768/5] calico.felix.fiptables 336: IptablesUpdater<queue_len=0,live=True,msg=None> Successfully processed iptables updates.
```
The latter should have replaced the `felix-p-_a0438770e0e8efc-[io]` stub chains with the correct versions.

Some ideas:
* this is during the graceful restart window so that could be causing a problem (but then restarting felix fixed it)
* there could be another app clobbering our iptables-restore
* there might have been a silent failure in the iptables-restore code.

@DaveLangridge could you give this a spin with debug logging enabled on the offending felix:
```
# /etc/calico/felix.cfg
LogSeverityFile = DEBUG
```
That will make the IptablesUpdater emit a lot more diags (including the actual text that it sends into iptables-restore).

**Comment by fasaxc** [comment]
@DaveLangridge @matthewdupre Any progress on this one?  There's certainly some more logs we could add to the IptablesUpdater.  Maybe a warning log whenever it programs a stub chain (one with ""WARNING..." in its comment).

**Comment by fasaxc** [comment]
(The counterpoint is that those chains are also present in the mainline case but they should be transient, only there for a second or so.)

**Comment by matthewdupre** [comment]
We don't believe this is necessarily a case of Felix programming anything wrong - it seems more likely that something else is clobbering IPTables and Felix never fixes it.

Having said that, we've been unable to reproduce, so we've shelved it until Dave gets back from holiday next week.

**Comment by MatMeredith** [comment]
I ran this test this morning (actually unaware of this issue).   The test ran (in that it eventually completed), but failed due to the following error logs from Felix at the point at which the VM on the power cycled compute host was restarted.   Are these of any interest?   

--------------------

=== LOG 2015-06-19 09:43:37 ERROR [pid 1819] ===
calico.felix.fiptables 618: iptables-restore failed with non-retryable error: ip(6)tables-restore failed with output: iptables-restore v1.4.21: Couldn't load target `felix-FORWARD':No such file or directory

Error occurred at line: 2
Try `iptables-restore -h' or 'iptables-restore --help' for more information.
.
Output:

Error:
iptables-restore v1.4.21: Couldn't load target `felix-FORWARD':No such file or directory

Error occurred at line: 2
Try `iptables-restore -h' or 'iptables-restore --help' for more information.

Input was:
*filter
--insert FORWARD --jump felix-FORWARD
COMMIT


=== LOG 2015-06-19 09:43:37 ERROR [pid 1819] ===
calico.felix.fiptables 618: iptables-restore failed with non-retryable error: ip(6)tables-restore failed with output: iptables-restore v1.4.21: Couldn't load target `felix-FORWARD':No such file or directory

Error occurred at line: 2
Try `iptables-restore -h' or 'iptables-restore --help' for more information.
.
Output:

Error:
iptables-restore v1.4.21: Couldn't load target `felix-FORWARD':No such file or directory

Error occurred at line: 2
Try `iptables-restore -h' or 'iptables-restore --help' for more information.

Input was:
*filter
--insert FORWARD --jump felix-FORWARD
COMMIT


=== LOG 2015-06-19 09:43:37 ERROR [pid 1819] ===
calico.felix.actor 233: Exception processing d450e8224e0b (ensure_rule_inserted)
Traceback (most recent call last):
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 231, in _step
result = msg.method()
File "/usr/lib/python2.7/dist-packages/calico/felix/fiptables.py", line 237, in ensure_rule_inserted
'COMMIT'])
File "/usr/lib/python2.7/dist-packages/calico/felix/fiptables.py", line 582, in _execute_iptables
futils.check_call(cmd, input_str=input_str)
File "/usr/lib/python2.7/dist-packages/calico/felix/futils.py", line 108, in check_call
args, retcode, stdout, stderr, input=input_str)
FailedSystemCall: Failed system call (retcode : 2, args : ('iptables-restore', '--noflush', '--verbose'))
stdout  :
stderr  : iptables-restore v1.4.21: Couldn't load target `felix-FORWARD':No such file or directory

Error occurred at line: 2
Try `iptables-restore -h' or 'iptables-restore --help' for more information.

input  : *filter
--insert FORWARD --jump felix-FORWARD
COMMIT



=== LOG 2015-06-19 09:43:37 ERROR [pid 1819] ===
calico.felix.actor 233: Exception processing d450e8224e0b (ensure_rule_inserted)
Traceback (most recent call last):
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 231, in _step
result = msg.method()
File "/usr/lib/python2.7/dist-packages/calico/felix/fiptables.py", line 237, in ensure_rule_inserted
'COMMIT'])
File "/usr/lib/python2.7/dist-packages/calico/felix/fiptables.py", line 582, in _execute_iptables
futils.check_call(cmd, input_str=input_str)
File "/usr/lib/python2.7/dist-packages/calico/felix/futils.py", line 108, in check_call
args, retcode, stdout, stderr, input=input_str)
FailedSystemCall: Failed system call (retcode : 2, args : ('iptables-restore', '--noflush', '--verbose'))
stdout  :
stderr  : iptables-restore v1.4.21: Couldn't load target `felix-FORWARD':No such file or directory

Error occurred at line: 2
Try `iptables-restore -h' or 'iptables-restore --help' for more information.

input  : *filter
--insert FORWARD --jump felix-FORWARD
COMMIT



=== LOG 2015-06-19 09:43:37 ERROR [pid 1819] ===
calico.felix.felix 180: Exception killing main greenlet
Traceback (most recent call last):
File "/usr/lib/python2.7/dist-packages/calico/felix/felix.py", line 148, in _main_greenlet
v4_nat_updater)
File "/usr/lib/python2.7/dist-packages/calico/felix/frules.py", line 159, in install_global_rules
async=False)
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 486, in queue_fn
blocking_result = result.get()
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 596, in get
timeout=timeout)
File "/usr/lib/python2.7/dist-packages/gevent/event.py", line 235, in get
raise self._exception
FailedSystemCall: Failed system call (retcode : 2, args : ('iptables-restore', '--noflush', '--verbose'))
stdout  :
stderr  : iptables-restore v1.4.21: Couldn't load target `felix-FORWARD':No such file or directory

Error occurred at line: 2
Try `iptables-restore -h' or 'iptables-restore --help' for more information.

input  : *filter
--insert FORWARD --jump felix-FORWARD
COMMIT



=== LOG 2015-06-19 09:43:37 ERROR [pid 1819] ===
calico.felix.felix 180: Exception killing main greenlet
Traceback (most recent call last):
File "/usr/lib/python2.7/dist-packages/calico/felix/felix.py", line 148, in _main_greenlet
v4_nat_updater)
File "/usr/lib/python2.7/dist-packages/calico/felix/frules.py", line 159, in install_global_rules
async=False)
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 486, in queue_fn
blocking_result = result.get()
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 596, in get
timeout=timeout)
File "/usr/lib/python2.7/dist-packages/gevent/event.py", line 235, in get
raise self._exception
FailedSystemCall: Failed system call (retcode : 2, args : ('iptables-restore', '--noflush', '--verbose'))
stdout  :
stderr  : iptables-restore v1.4.21: Couldn't load target `felix-FORWARD':No such file or directory

Error occurred at line: 2
Try `iptables-restore -h' or 'iptables-restore --help' for more information.

input  : *filter
--insert FORWARD --jump felix-FORWARD
COMMIT

2015-06-19 09:45:37,009 [ERROR][7528] calicotest.openstack.basetest 127: ****** End of errors in logs from /var/log/calico/felix.log on calico-cidev13.datcon.co.uk ******
ERROR

**Comment by MatMeredith** [comment]
Another run, another different set of Felix errors following power cycle of the compute node.   Again it seems that everything recovered this time though -- new VMs were able to be created and connectivity was confirmed to be correct.

2015-06-19 12:25:02,518 [ERROR][10998] calicotest.openstack.basetest 125: ***** Start of errors in logs from /var/log/calico/felix.log on calico-cidev13.datcon.co.uk *****

=== LOG 2015-06-19 12:23:39 ERROR [pid 1809] ===
calico.felix.fiptables 358: Some of our chains disappeared from the dataplane: set(['felix-INPUT', 'felix-FORWARD']). Raising an exception.

=== LOG 2015-06-19 12:23:39 ERROR [pid 1809] ===
calico.felix.fiptables 358: Some of our chains disappeared from the dataplane: set(['felix-INPUT', 'felix-FORWARD']). Raising an exception.

=== LOG 2015-06-19 12:23:39 ERROR [pid 1809] ===
calico.felix.actor 233: Exception processing 67ac723c9b16 (cleanup)
Traceback (most recent call last):
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 231, in _step
result = msg.method()
File "/usr/lib/python2.7/dist-packages/calico/felix/fiptables.py", line 360, in cleanup
"Felix chains missing from iptables: %s" % missing_chains)
IptablesInconsistent: Felix chains missing from iptables: set(['felix-INPUT', 'felix-FORWARD'])

=== LOG 2015-06-19 12:23:39 ERROR [pid 1809] ===
calico.felix.actor 233: Exception processing 67ac723c9b16 (cleanup)
Traceback (most recent call last):
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 231, in _step
result = msg.method()
File "/usr/lib/python2.7/dist-packages/calico/felix/fiptables.py", line 360, in cleanup
"Felix chains missing from iptables: %s" % missing_chains)
IptablesInconsistent: Felix chains missing from iptables: set(['felix-INPUT', 'felix-FORWARD'])

=== LOG 2015-06-19 12:23:39 ERROR [pid 1809] ===
calico.felix.actor 233: Exception processing da3d26898d87 (trigger_cleanup)
Traceback (most recent call last):
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 231, in _step
result = msg.method()
File "/usr/lib/python2.7/dist-packages/calico/felix/splitter.py", line 110, in trigger_cleanup
ipt_updater.cleanup(async=False)
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 486, in queue_fn
blocking_result = result.get()
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 596, in get
timeout=timeout)
File "/usr/lib/python2.7/dist-packages/gevent/event.py", line 235, in get
raise self._exception
IptablesInconsistent: Felix chains missing from iptables: set(['felix-INPUT', 'felix-FORWARD'])

=== LOG 2015-06-19 12:23:39 ERROR [pid 1809] ===
calico.felix.actor 233: Exception processing da3d26898d87 (trigger_cleanup)
Traceback (most recent call last):
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 231, in _step
result = msg.method()
File "/usr/lib/python2.7/dist-packages/calico/felix/splitter.py", line 110, in trigger_cleanup
ipt_updater.cleanup(async=False)
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 486, in queue_fn
blocking_result = result.get()
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 596, in get
timeout=timeout)
File "/usr/lib/python2.7/dist-packages/gevent/event.py", line 235, in get
raise self._exception
IptablesInconsistent: Felix chains missing from iptables: set(['felix-INPUT', 'felix-FORWARD'])


2015-06-19 12:25:02,519 [ERROR][10998] calicotest.openstack.basetest 127: ****** End of errors in logs from /var/log/calico/felix.log on calico-cidev13.datcon.co.uk ******
ERROR

**Comment by fasaxc** [comment]
@MatMeredith That looks like some other process clobbered our iptables rules.  Felix detected the inconsistency and (should have) restarted, fixing the problem.

**Comment by fasaxc** [comment]
(Or a bug in felix...)

**Comment by fasaxc** [comment]
@MatMeredith Just spotted your first error report too.  That one looks like someone deleted our chain out from under us too.  That error should trigger felix to restart and fix the chain.

@DaveLangridge I wonder if the host has a job that restores iptables state at boot and it's running concurrently with felix and clobbering our programming.

**Comment by fasaxc** [comment]
@MatMeredith Dug into this a bit more.  iptables-persistent could be the problem in your case, which appears different to @DaveLangridge's.  iptables-persistent runs at bootup and flushes the chains and then fills them with the contents of `/etc/iptables/rules.v[46]`.  I think iptables-persistent logs to syslog when it runs so you could confirm by examining syslog and seeing if it appears to start concurrenlty/after felix.

**Comment by fasaxc** [comment]
Scratch that, it only logs to the boot log, which doesn't have timestamps to correlate.

**Comment by DaveLangridge** [comment]
Has the behavior here improved? Now when I run the test it succeeded, but I hit some exceptions, and then it recovered:
2015-06-26 14:05:24,336 [ERROR][1530/6] calico.felix.fiptables 363: Some of our chains disappeared from the dataplane: set(['felix-p-_fa54ada24ca3dc7-i', 'felix-p-_fa54ada24ca3dc7-o', u'felix-from-c2cd573e-a4', u'felix-to-c2cd573e-a4']). Raising an exception.
2015-06-26 14:05:24,336 [ERROR][1530/6] calico.felix.fiptables 363: Some of our chains disappeared from the dataplane: set(['felix-p-_fa54ada24ca3dc7-i', 'felix-p-_fa54ada24ca3dc7-o', u'felix-from-c2cd573e-a4', u'felix-to-c2cd573e-a4']). Raising an exception.
2015-06-26 14:05:24,346 [ERROR][1530/6] calico.felix.actor 233: Exception processing 507dfb2e2d24 (cleanup)
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 231, in _step
    result = msg.method()
  File "/usr/lib/python2.7/dist-packages/calico/felix/fiptables.py", line 365, in cleanup
    "Felix chains missing from iptables: %s" % missing_chains)
IptablesInconsistent: Felix chains missing from iptables: set(['felix-p-_fa54ada24ca3dc7-i', 'felix-p-_fa54ada24ca3dc7-o', u'felix-from-c2cd573e-a4', u'felix-to-c2cd573e-a4'])
2015-06-26 14:05:24,346 [ERROR][1530/6] calico.felix.actor 233: Exception processing 507dfb2e2d24 (cleanup)
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 231, in _step
    result = msg.method()
  File "/usr/lib/python2.7/dist-packages/calico/felix/fiptables.py", line 365, in cleanup
    "Felix chains missing from iptables: %s" % missing_chains)
IptablesInconsistent: Felix chains missing from iptables: set(['felix-p-_fa54ada24ca3dc7-i', 'felix-p-_fa54ada24ca3dc7-o', u'felix-from-c2cd573e-a4', u'felix-to-c2cd573e-a4'])
2015-06-26 14:05:24,347 [INFO][1530/6] calico.felix.fiptables 380: IptablesUpdater<queue_len=0,live=True,msg=None> no updates in this batch.
2015-06-26 14:05:24,348 [ERROR][1530/7] calico.felix.actor 233: Exception processing 874f91e9ed3d (trigger_cleanup)
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 231, in _step
    result = msg.method()
  File "/usr/lib/python2.7/dist-packages/calico/felix/splitter.py", line 110, in trigger_cleanup
    ipt_updater.cleanup(async=False)
  File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 486, in queue_fn
    blocking_result = result.get()
  File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 596, in get
    timeout=timeout)
  File "/usr/lib/python2.7/dist-packages/gevent/event.py", line 235, in get
    raise self._exception
IptablesInconsistent: Felix chains missing from iptables: set(['felix-p-_fa54ada24ca3dc7-i', 'felix-p-_fa54ada24ca3dc7-o', u'felix-from-c2cd573e-a4', u'felix-to-c2cd573e-a4'])
2015-06-26 14:05:24,348 [ERROR][1530/7] calico.felix.actor 233: Exception processing 874f91e9ed3d (trigger_cleanup)
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 231, in _step
    result = msg.method()
  File "/usr/lib/python2.7/dist-packages/calico/felix/splitter.py", line 110, in trigger_cleanup
    ipt_updater.cleanup(async=False)
  File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 486, in queue_fn
    blocking_result = result.get()
  File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 596, in get
    timeout=timeout)
  File "/usr/lib/python2.7/dist-packages/gevent/event.py", line 235, in get
    raise self._exception
IptablesInconsistent: Felix chains missing from iptables: set(['felix-p-_fa54ada24ca3dc7-i', 'felix-p-_fa54ada24ca3dc7-o', u'felix-from-c2cd573e-a4', u'felix-to-c2cd573e-a4'])
2015-06-26 14:05:24,459 [INFO][2080/1] calico.common 286: Logging initialized

**Comment by DaveLangridge** [comment]
Just noticed that this set of logs is the same as one of @MatMeredith 's above.

**Comment by fasaxc** [comment]
@DaveLangridge If felix's chains are being clobbered, then the behaviour will depend on when that happens:

* Very early and it'll clobber the the from-ENDPOINT chain before we insert the rule into the FORWARD chain
* A bit later and we'll catch it at cleanup time.
* After that, we're unlikely to spot it.

**Comment by fasaxc** [comment]
I had a play at reproing this locally, trying to see it if was iptables-persistent causing the issue.  I couldn't seem to get it to break on Ubuntu.  iptables-persistent always runs before felix starts.

To verify that, I actually hacked the iptables-persistent script to print before it started applying the updates.  Also verified the upstart job for felix starts on the correct event to make that the case.

**Comment by DaveLangridge** [comment]
I've seen the two sets of logs above, plus the case where felix doesn't spot it. I assume these are your 3 cases. We don't see this problem at all for the equivalent reboot test. Only for power cycle.

**Comment by fasaxc** [comment]
@DaveLangridge Looks like there are a couple of likely contenders on the system that shows this problem.  iptables-persistent is running but I believe that always runs before calico-felix.  There are also chains with prefix "nova-api-metadat-...", which are not present in `/etc/iptables/rules.v4` so those must be inserted by another component.

**Comment by Lukasa** [comment]
nova-api-metadata is the culprit there. ;)

**Comment by fasaxc** [comment]
Yep , just digging through its code.  It does an `iptables-save`, makes its changes in-memory and then does an `iptables-restore` with the whole lot.  That's vulnerable to a race with anything else that is modifying the table concurrently.

**Comment by matthewdupre** [comment]
Talked with @fasaxc - couple of options so far:
- upstream fix to nova-metadata-api
- don't run the nova-metadata-api on compute hosts - NAT metadata to the controller instead.

**Comment by matthewdupre** [comment]
We can at least try the second.

**Comment by DaveLangridge** [comment]
How about a defensive fix in felix? Checking config at intervals (like the plugin resynch). Then we'd recover from this, and any other cases where someone screwed up the iptables.

**Comment by matthewdupre** [comment]
@DaveLangridge we've already done that - see #687.

**Comment by DaveLangridge** [comment]
So we have. I'm not sure if I care about any other fixes. We've only seen this on power cycle, an I suspect there will be all sorts of stuff hitting the fan at that time. Will anyone notice if it takes a bit longer to get up and running? Main concern before was we could get into this state and not recover.

**Comment by matthewdupre** [comment]
We've discussed this a bit.  There are a few ways forward, but as @DaveLangridge says we've already made a partial fix for this that might be good enough.

1. Don't run nova-api-metadata.  Doesn't help us if other people/processes modify IPTables badly, doesn't play properly with combined controllers + computes (where nova-api does the same, and we can't not run that) and increases load on controllers.

2. Improve nova-api-metadata.  Doesn't help us if other people/processes modify IPTables badly.

3. More frequent Felix resyncs.  Possible performance problem - particularly because it will put heavy load on etcd.  We could work around this by adding some facilty for smaller resyncs that just fix up IPTables without requerying etcd.

**Comment by fasaxc** [comment]
Pushing this out to 0.28.  I think we understand what's going on now but we haven't even started a fix.

**Comment by fasaxc** [comment]
Work in progress on https://github.com/Metaswitch/calico/tree/smc-ipt-refresh

**Comment by plwhite** [comment]
OK, so I would prefer that Felix does not need to know the state of all iptables chains. I think there are several options.

Let's dump the option of Felix could just doing a total resync much more frequently. This is probably too expensive in terms of CPU and hit on etcd.

1. Felix could remember its iptables state, and apply it periodically. Thus every 30 seconds or so it could go check that things are as expected. This is ugly (too much occupancy, and also we still have the 30 second problem). However, it plausibly could work.

2. We could fix the code that is nuking our iptables chains under our feet (i.e. fix the misbehaving application). I think it's not unreasonable that we don't work if there's a broken application, although since it's nova we probably do need to do something about it!

3. We could have a halfway house. Nothing should be nuking the Felix chains, and any code that does is broken (and we should fix it). However, we could argue that we should then

    * fix up the nova code so that it only alters the chains it cares about;
    * extend Felix so that it frequently validates only the starting chains (INPUT, FORWARD - the ones it cares about).

4. We could do something that is frankly a hack - restart Felix after nova completes doing its thing.

I prefer option 3 over 2, and option 2 over 1. But I have no feel for the cost of the changes to nova code, and I also have a concern that getting fixes into nova is painful. Anybody like to comment on this? Option 4 is nasty, but perhaps worth considering if we need to buy time / get a cheap fix out quickly.

**Comment by plwhite** [comment]
We've agreed pragmatically to do 1 and 2 - but with 2 following later.

**Comment by eepyaich** [comment]
@plwhite @fasaxc Have we done the first thing here?  Can we close this issue and have a separate one to track any long term follow up?
